jetty-core/jetty-http2/jetty-http2-client-transport/src/main/java/org/eclipse/jetty/http2/client/transport/internal/HttpReceiverOverHTTP2.java

```diff
@@ -93,46 +130,31 @@ private void onResponse(Stream stream, HeadersFrame frame)
 
-        if (responseBegin(exchange))
+        responseBegin(exchange);
+        HttpFields headers = response.getFields();
+        for (HttpField header : headers)
         {
-            HttpFields headers = response.getFields();
-            for (HttpField header : headers)
-            {
-                if (!responseHeader(exchange, header))
-                    return;
-            }
-
-            HttpRequest httpRequest = exchange.getRequest();
-            if (MetaData.isTunnel(httpRequest.getMethod(), httpResponse.getStatus()))
-            {
-                ClientHTTP2StreamEndPoint endPoint = new ClientHTTP2StreamEndPoint((HTTP2Stream)stream);
-                long idleTimeout = httpRequest.getIdleTimeout();
-                if (idleTimeout > 0)
-                    endPoint.setIdleTimeout(idleTimeout);
-                if (LOG.isDebugEnabled())
-                    LOG.debug("Successful HTTP2 tunnel on {} via {}", stream, endPoint);
-                ((HTTP2Stream)stream).setAttachment(endPoint);
-                HttpConversation conversation = httpRequest.getConversation();
-                conversation.setAttribute(EndPoint.class.getName(), endPoint);
-                HttpUpgrader upgrader = (HttpUpgrader)conversation.getAttribute(HttpUpgrader.class.getName());
-                if (upgrader != null)
-                    upgrade(upgrader, httpResponse, endPoint);
-            }
+            responseHeader(exchange, header);
+        }
 
-            if (responseHeaders(exchange))
-            {
-                int status = response.getStatus();
-                if (frame.isEndStream() || HttpStatus.isInterim(status))
-                    responseSuccess(exchange);
-                else
-                    stream.demand();
-            }
-            else
-            {
-                if (LOG.isDebugEnabled())
-                    LOG.debug("Stalling response processing, no demand after headers on {}", this);
-            }
+        HttpRequest httpRequest = exchange.getRequest();
+        if (MetaData.isTunnel(httpRequest.getMethod(), httpResponse.getStatus()))
+        {
+            ClientHTTP2StreamEndPoint endPoint = new ClientHTTP2StreamEndPoint((HTTP2Stream)stream);
+            long idleTimeout = httpRequest.getIdleTimeout();
+            if (idleTimeout > 0)
+                endPoint.setIdleTimeout(idleTimeout);
+            if (LOG.isDebugEnabled())
+                LOG.debug("Successful HTTP2 tunnel on {} via {}", stream, endPoint);
+            ((HTTP2Stream)stream).setAttachment(endPoint);
+            HttpConversation conversation = httpRequest.getConversation();
+            conversation.setAttribute(EndPoint.class.getName(), endPoint);
+            HttpUpgrader upgrader = (HttpUpgrader)conversation.getAttribute(HttpUpgrader.class.getName());
+            if (upgrader != null)
+                upgrade(upgrader, httpResponse, endPoint);
         }
+
+        responseHeaders(exchange);
```

Comment:
Can we pass a boolean to indicate whether the HEADERS frame is the last, so we skip notifying the content?